src/Ocelot/Multiplexer/MultiplexingMiddleware.cs

```diff
@@ -1,219 +1,265 @@
-                MapNotAggregate(httpContext, singleResponse);
-                return;
-            }
+        // case 3: if multiple downstream routes
+        var routeKeysConfigs = route.DownstreamRouteConfig;
+        if (routeKeysConfigs == null || routeKeysConfigs.Count == 0)
+        {
+            await ProcessRoutes(httpContext, route);
+            return;
+        }
 
-            if (route.DownstreamRouteConfig?.Any() != true)
-            {
-                var tasks = new Task<HttpContext>[route.DownstreamRoute.Count];
+        // case 4: if multiple downstream routes with route keys
+        var mainResponse = await ProcessMainRoute(httpContext, downstreamRoutes[0]);
+        if (mainResponse == null)
+        {
+            return;
+        }
 
-                for (var i = 0; i < route.DownstreamRoute.Count; i++)
-                {
-                    var newHttpContext = Copy(httpContext);
+        var tasksList = await ProcessRoutesWithRouteKeys(httpContext, downstreamRoutes, routeKeysConfigs, mainResponse);
+        if (tasksList.Length == 0)
+        {
+            return;
+        }
 
-                    newHttpContext.Items
-                        .Add("RequestId", httpContext.Items["RequestId"]);
-                    newHttpContext.Items
-                        .SetIInternalConfiguration(httpContext.Items.IInternalConfiguration());
-                    newHttpContext.Items
-                        .UpsertTemplatePlaceholderNameAndValues(httpContext.Items.TemplatePlaceholderNameAndValues());
-                    newHttpContext.Items
-                        .UpsertDownstreamRoute(route.DownstreamRoute[i]);
+        await MapResponses(httpContext, route, mainResponse, tasksList.ToList());
+    }
 
-                    tasks[i] = Fire(newHttpContext, _next);
-                }
+    /// <summary>
+    /// Helper method to determine if only the first downstream route should be processed.
+    /// It is the case if the request is a websocket request or if there is only one downstream route.
+    /// </summary>
+    /// <param name="context">The http context.</param>
+    /// <param name="routes">The downstream routes.</param>
+    /// <returns>True if only the first downstream route should be processed.</returns>
+    private static bool ShouldProcessSingleRoute(HttpContext context, ICollection routes) => context.WebSockets.IsWebSocketRequest || routes.Count == 1;
+
+    /// <summary>
+    /// Processing a single downstream route (no route keys).
+    /// In that case, no need to make copies of the http context.
+    /// </summary>
+    /// <param name="context">The http context.</param>
+    /// <param name="route">The downstream route.</param>
+    /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+    protected virtual Task ProcessSingleRoute(HttpContext context, DownstreamRoute route)
+    {
+        context.Items.UpsertDownstreamRoute(route);
+        return _next.Invoke(context);
```

Comment:
I like it! 👍  Async method but without awaiting!
Not an issue.

Giu, it is `virtual` ❗ 😱 Why?

Reply:
It's protected virtual. That was unfortunately the only way to test the middleware properly. But, it's protected, so I thought, it's fine. You know that I don't like virtuals...

Reply:
I still don't like virtual... I just cheated with my own morals 😭

Reply:
Hush!... But you like protected mocks and `virtual` now ! Yeah!!! 🤠